src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,292 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <uv.h>
+
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/lua/executor.h"
+#include "nvim/os/keyboard.h"
+#include "nvim/vim.h"
+#include "nvim/version.h"
+
+#if defined(__linux__)
+# define XLIB_NAME "libX11.so.6"
+
+typedef void *_XDisplay;
+
+/// Holds the data of the present state of Keyboard
+typedef struct {
+  char data[18];
+} _XkbStateRec;
+typedef int _XStatus;
+
+/// The offset of the keyboard lock modes for X11
+enum { kLockedModsOffset = 9 };
+enum { _XkbUseCoreKbd = 0x0100 };
+
+/// Masks used in libX11 for different Lock states
+typedef enum {
+    kXNumLock = 0x10,  ///< Mask for NumLock
+    kXCapsLock = 0x02,  ///< Mask for CapsLock
+    kXScrollLock = 0x80,  ///< Mask for ScrollLock
+} XlibKbdLocks;
+#elif defined(WIN32)
+# include <Winuser.h>
+#elif defined(__APPLE__)
+# define Boolean Boolean_I_Dont_Care
+# include <IOKit/IOKitLib.h>
+# include <IOKit/IOReturn.h>
+# include <IOKit/hidsystem/IOHIDLib.h>
+# include <IOKit/hidsystem/IOHIDParameter.h>
+# include <CoreFoundation/CoreFoundation.h>
+# undef Boolean
+#endif
+
+/// Masks for the specific keyboard lock status
+typedef enum {
+  kNumLock = 0x01,  ///< Mask used to check if the NumLock is active
+  kCapsLock = 0x02,  ///< Mask used to check if the CapsLock is active
+  kScrollLock = 0x04,  ///< Mask used to check if the ScrollLock is active
+} KbdLocks;
+
+/// Used as a bitmask for the status of the different Lock Status
+typedef int ModMask;
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/keyboard.c.generated.h"
+#endif
+
+#if defined(__linux__)
+static bool xlib_opened = false;
+static uv_lib_t xlib;
+static _XDisplay *(*_XOpenDisplay)(void *);
+static int (*_XCloseDisplay)(_XDisplay *);
+static _XStatus (*_XkbGetState)(_XDisplay *, unsigned, _XkbStateRec *);
+
+/// A helper function to be used to load a function of libX11
+static bool os_xlib_dlsym(const char *name, void **fun, Error *err)
+  FUNC_ATTR_NONNULL_ALL
+{
+  const int ret = uv_dlsym(&xlib, name, fun);
+  if (ret != 0) {
+    api_set_error(err, kErrorTypeException,
+                  "Unable to obtain %s function from X11",
+                  name);
+    return false;
+  }
+  return true;
+}
+
+/// A helper function to be used to load the libX11 and required functions
```

Comment:
This also missing documentation about return value. And do not add noise about something being a “helper function”, most functions are like this.

```C
/// Initialize global variables, loading X11 library and required functions
///
/// @param[out]  err  Location where error is saved.
///
/// @return true if everything was loaded successfully, false otherwise.
```

(and do not use “helper function” here, there is no clear definition of “helper” for this statement to be useful, but it does occupy place)

Reply:
OK